### PR TITLE
Finished excercise 3

### PR DESCRIPTION
**Pull request by AnaCDomingo**

### CODE REVIEW

File: 01-JS-DOM-APIs/Ex03/index.html [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+   		<h1>This it my third JS excercise!</h1>
+   	</header>
+   	<section id="jokeTextbox"></section> <br>
+      <button onclick="getJoke('http://api.icndb.com/jokes/random')"> I have a joke for you </button>
+      <script>
+         function getJoke(url) {
+            var joke;
+            var jokeRequest = new XMLHttpRequest();
+            jokeRequest.onreadystatechange = function() {
+            if (this.readyState == 4 && this.status == 200) {
```
Comment (mtulian) [inline 01-JS-DOM-APIs/Ex03/index.html]: At first look, I thought that `jokeRequest.open("Get", url, true)` and the next sentence were outside the function, which is no posible. Try to make the code clearer, for example:

```
let getJoke = function (url) {
    let joke; //We should use es6
    let jokeRequest = new XMLHttpRequest();
   
    jokeRequest.onreadystatechange = function () {
    
        if (this.readyState === 4 && this.status === 200) { //Take care about the equal comparision, I recommend you to use the deep comparision instead
             joke = JSON.parse(this.responseText);
             document.getElementById('jokeTextbox').innerHTML = joke.value.joke; //Always use single quoutes for javascript code. The double quotes are for html
         }
    }
   
    jokeRequest.open('GET', url, true);
    jokeRequest.send();
}
```

**Review by mtulian — COMMENTED** [review]
It's fine, just take a look into the advice and let me know what you thing about

File: 01-JS-DOM-APIs/Ex03/index.html (hunk 1 above)
Comment (AnaCDomingo) [reply 01-JS-DOM-APIs/Ex03/index.html]: I'm learning JavaScript from the beginning, and I found your advices so useful... Sometimes I think I'm dumb, but I think is part of the process! I'll correct these things.

File: 01-JS-DOM-APIs/Ex03/index.html [hunk 2]
```diff
@@ -1,33 +1,45 @@
...
+    <link rel="stylesheet" href="css/styles.css">
+    <link href="https://fonts.googleapis.com/css?family=Bubbler+One" rel="stylesheet">
+  </head>
+  <body>
+    <header>
+     	<h1>This it my third JS excercise!</h1>
+    </header>
+    <section id="jokeTextbox"></section> <br>
+    <button onclick="getJoke (jconfig)"> I have a joke for you </button>
+    <script>
```
Comment (mtulian) [inline 01-JS-DOM-APIs/Ex03/index.html]: In the first two exercises were fine, but let's start with the creation of new files for javascript.

File: 01-JS-DOM-APIs/Ex03/index.html [hunk 3]
```diff
@@ -1,33 +1,45 @@
...
-               document.getElementById("jokeTextbox").innerHTML = joke.value.joke;
+    <link rel="stylesheet" href="css/styles.css">
+    <link href="https://fonts.googleapis.com/css?family=Bubbler+One" rel="stylesheet">
+  </head>
+  <body>
+    <header>
+     	<h1>This it my third JS excercise!</h1>
+    </header>
+    <section id="jokeTextbox"></section> <br>
+    <button onclick="getJoke (jconfig)"> I have a joke for you </button>
```
Comment (mtulian) [inline 01-JS-DOM-APIs/Ex03/index.html]: Don't need to provide the jsconfig here. If you already have it available in javascript, let's consume it from there.

File: 01-JS-DOM-APIs/Ex03/index.html [hunk 4]
```diff
@@ -1,33 +1,45 @@
 <!DOCTYPE html>
 <html>
-   <head>
+  <head>
 		<meta charset="UTF-8">
 		<meta http-equiv="x-ua-compatible" content="ie=edge">
```
Comment (mtulian) [inline 01-JS-DOM-APIs/Ex03/index.html]: **MInor**: Review de identation here

File: 01-JS-DOM-APIs/Ex03/index.html [hunk 5]
```diff
@@ -1,33 +1,45 @@
...
+    <link href="https://fonts.googleapis.com/css?family=Bubbler+One" rel="stylesheet">
+  </head>
+  <body>
+    <header>
+     	<h1>This it my third JS excercise!</h1>
+    </header>
+    <section id="jokeTextbox"></section> <br>
+    <button onclick="getJoke (jconfig)"> I have a joke for you </button>
+    <script>
+      var jconfig = {
```
Comment (mtulian) [inline 01-JS-DOM-APIs/Ex03/index.html]: Did you use it the `var` here for a particular reason?

File: 01-JS-DOM-APIs/Ex03/index.html [hunk 6]
```diff
@@ -1,33 +1,45 @@
...
+        fmethod:'GET',
+        url:'http://api.icndb.com/jokes/random',
+        sync:true
+      }
+      function getJoke (config) {
+        let promise = new Promise (function(resolve, reject) {
+          let jokeRequest = new XMLHttpRequest();
+          jokeRequest.open(config.fmethod, config.url, config.sync);
```
Comment (mtulian) [inline 01-JS-DOM-APIs/Ex03/index.html]: If I see `getJoke` as a black box, the input and the output are the expected. Let's analyze the requirement together:

First, create a function that returns a joke, 

```
let getJoke = function () {
     //call reusable function
    
     return joke;
}

```
Then, create a reusable function that perfroms AJAX calls. That function will receive a config object and return a promise.

```
let performAjaxCall = function (config) {
     return new Promise(this.promiseHandler); // the promise should be 
}

let promiseHandler = function (resolve, reject) {
     //We should manipulate the promise here, determine when we should resolve or reject it
};
```

Give a second try and let me know about the results.

File: 01-JS-DOM-APIs/Ex03/script.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  return new reqConfig ("GET", "http://api.icndb.com/jokes/random", true);
+}
+function reRequest(object) {
+  let req;
+  return new Promise (function(resolve, reject) {
+    req = new XMLHttpRequest();
+    req.open(object.method, object.url, object.sync);
+    req.send();
+    req.onload = function () {
+      if (this.status === 200 && this.readyState === 4) {
```
Comment (mtulian) [inline 01-JS-DOM-APIs/Ex03/script.js:16]: Not need to check the readyState, once onload is executed the readyState is equals to 4.

File: 01-JS-DOM-APIs/Ex03/script.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        resolve (this.response);
```
Comment (mtulian) [inline 01-JS-DOM-APIs/Ex03/script.js:17]: The response has the request response?

**Review by mtulian — APPROVED** [review]
It's fine, go a head a merge it!

File: 01-JS-DOM-APIs/Ex03/script.js (hunk 7 above)
Comment (AnaCDomingo) [reply 01-JS-DOM-APIs/Ex03/script.js:16]: What if I use "onReadyStateChange"?

File: 01-JS-DOM-APIs/Ex03/script.js (hunk 8 above)
Comment (AnaCDomingo) [reply 01-JS-DOM-APIs/Ex03/script.js:17]: That's the idea :grimacing:

**Pull request merged**